pyesbulk/__init__.py

```diff
@@ -81,14 +126,72 @@ def quiet_loggers():
     A convenience function to quiet the urllib3 and elasticsearch1 loggers.
     """
     logging.getLogger("urllib3").setLevel(logging.FATAL)
-    logging.getLogger(_es_logger).setLevel(logging.FATAL)
+    logging.getLogger("elasticsearch").setLevel(logging.FATAL)
 
 
 # The 5xx codes on which template PUT operations are retried.
 _RETRY_5xxS = [500, 503, 504]
 
 
-def put_template(es, name=None, mapping_name=None, body=None):
+def _get_meta_version(name, body):
+    """
+    _get_meta_version Try to find an "_meta":{"version": "value"} in the
+    specified template body.
+
+    For V7 and beyond, a template should not have a name, which means the
+    mapping body includes an "_meta" key directly.
+
+    For V6 and V7 there may be an "_doc" key in the top level dict, which
+    will have "_meta" as a second level key.
+
+    Earlier versions will normally have a "document type" as the first
+    level key, with "_meta" underneath.
+
+    We're being accommodating here: if the first level dict includes an
+    "_meta" key, we'll use it. If not, and the first level dict has a single
+    key, we'll look for "_meta" under that.
+
+    Args:
+        body (dict): An Elasticsearch index document template "mappings"
+        document.
+    """
+    if not isinstance(body, dict):
+        raise TypeError
+    underbody = body
+    if "_meta" not in body:
+        # Elasticsearch V1 allows multiple named templates, but since they're
+        # going into the same index we require that the versions be the same
+        # since we can return only one.
+        underbody = body[next(iter(body))]
+        version = None
+        for key in body:
+            try:
+                v = int(body[key]["_meta"]["version"])
+                if not version:
+                    version = v
+                else:
+                    if v != version:
+                        raise Exception(
+                            f"Bad template, multiple templates with "
+                            "differing versions at {key}"
+                        )
+            except KeyError:
+                raise Exception(
+                    f"Bad template, {key}: template has no '_meta' version"
+                )
```

Comment:
When the `for` loop at line 167 completes without raising any exceptions, `version` will be assigned a value.  But then the `try` block at line 182 will execute next, over-writing the value in `version`.  Shouldn't we be returning `version` at the end of the `if` block of line 167?

Reply:
Indeed; I didn't think this through totally when I added the multi-document validation loop. Also, the loop dereference isn't inside a `try` so we could get caught. Well, OK then...

Reply:
Shoot; and I pushed a new commit without addressing this. Darn.